Fix comment parsing inside command substitutions and brackets

shayaviv opened the pull request:
This change avoids parsing brackets and quotes within comments.

This fixes the following examples (as well as those described in #7866 and #8022):
| Script | Before | After |
| - | - | - |
| <pre lang="fish">echo a (&#xA;    echo b  # I'm using an apostrophe &#xA;)</pre> | (line 1): Unexpected end of string, quotes are not balanced | `a b` |
| <pre lang="fish">echo a (&#xA;    echo b  # Oops, closing paren :)&#xA;)</pre> | (line 3): Unexpected ')' for unopened parenthesis | `a b` |
| <pre lang="fish">echo a[b #c]</pre> | `a[b #c]` | Unexpected end of string, square brackets do not match |

Fixes #7866, and fixes #8022.

## TODOs:
- [x] Changes to fish usage are reflected in user documentation/manpages.
  - _There are no changes in usage._
- [x] Tests have been added for regressions fixed
- [x] User-visible changes noted in CHANGELOG.rst

floam commented:
Well I'm a fan, nice job, cool!

The only thing I'd want to think on is if this is the simplest, best fix. These are parts of the codebase that are super convoluted.

shayaviv commented:
I refactored my changes - they are now the simplest I could get them without touching too much of the existing parsing code. I also added regression tests to `basic.fish`.

krobelus commented:
Looks pretty nice! :+1:

> Should we note this change?

I think so.

krobelus commented on src/parse_util.cpp:
```diff
@@ -167,6 +170,7 @@ static int parse_util_locate_cmdsub(const wchar_t *in, const wchar_t **begin, co
         } else {
             escaped = false;
         }
+        is_token_begin = !escaped && (iswspace(*pos) || *pos == L'(');
```
This criteria is missing some other characters that separate tokens, like `;`.
`tok_is_string_character()` should give the complete answer (the API is a bit awkward, sorry).

(Using that, we also might be able to avoid the `- 1` in `comment_end(pos) - 1`)

shayaviv replied:
Thank, but I believe `;` does not separate comment tokens. Here's an example:

```fish
echo a  # ; echo b
```

According to the existing tokenizer.cpp code that finds the end of a comment token (which I extracted to `comment_end()`), the comment token should span all the way to a `L'\n'` or `L'\0'`: https://github.com/fish-shell/fish-shell/blob/964b7a729a7ea50485b9ce56861f5e922ac26a2a/src/tokenizer.cpp#L540-L545

As for the `- 1`, it's there because at the end of the loop iteration, `pos` is incremented (`pos++`) before being checked (`*pos != L'\0'`). Without the `- 1`, `pos` is incremented beyond a null delimiter, and then `*pos` is an an out-of-bounds read.

krobelus replied:
I meant something else: a # will also start a token in `echo foo;#bar`, hence the check is to narrow

example test case:

```fish
echo (;#"
      echo baz)
# CHECK: baz
```

shayaviv replied:
I see - thanks for catching that! I'll fix that.

krobelus commented on tests/checks/basic.fish:
```diff
@@ -533,6 +533,17 @@ echo banana
+# CHECK: bar
+echo (#"
+      echo baz)
+# CHECK: baz
```
I think you're also fixing #8022, so how about adding

```fish
# Make sure we don't match up brackets within comments (#8022).
$fish -c 'echo f[oo # not valid, no matching ]'
# CHECKERR: fish: Unexpected end of string, square brackets do not match
# CHECKERR: echo f[oo # not valid, no matching ]
# CHECKERR: {{      }}^
```

krobelus commented:
Nice work! I have found one minor issue

Also, consider squashing the commits, so the changes are easier to follow for third parties (including
future you and future me).

krobelus commented on src/tokenizer.cpp:
```diff
@@ -679,6 +682,10 @@ maybe_t<tok_t> tokenizer_t::next() {
 }
 
+bool is_token_delimiter(wchar_t c, bool is_first, maybe_t<wchar_t> next) {
+    return c == L'(' || !tok_is_string_character(c, is_first, next);
```
+bool is_token_delimiter(wchar_t c, bool is_first, maybe_t<wchar_t> next) {
+    return c == L'(' || !tok_is_string_character(c, is_first, next);

Interesting. This function deviates from tok_is_string_character() because today `(a)(b)(c)` is a single token at parsing time. Only during execution, command substitutions are reparsed and further split.  We could change this in future.

Here is an alternative solution I played around with, but I think we should just discard it, because
`is_comment()` doesn't deal with backslash escapes itself but requires the caller to do so, which is weird.

<details>

```diff
diff --git a/src/parse_util.cpp b/src/parse_util.cpp
index c626804d0..82870c6d9 100644
--- a/src/parse_util.cpp
+++ b/src/parse_util.cpp
@@ -90,8 +90,6 @@ size_t parse_util_get_offset(const wcstring &str, int line, long line_offset) {
 static int parse_util_locate_cmdsub(const wchar_t *in, const wchar_t **begin, const wchar_t **end,
                                     bool allow_incomplete, bool *inout_is_quoted) {
     bool escaped = false;
-    bool is_first = true;
-    bool is_token_begin = true;
     bool syntax_error = false;
     int paran_count = 0;
     std::vector<int> quoted_cmdsubs;
@@ -126,7 +124,7 @@ static int parse_util_locate_cmdsub(const wchar_t *in, const wchar_t **begin, co
                 if (!process_opening_quote(*pos)) break;
             } else if (*pos == L'\\') {
                 escaped = true;
-            } else if (*pos == L'#' && is_token_begin) {
+            } else if (is_comment(pos, in)) {
                 pos = comment_end(pos) - 1;
             } else {
                 if (*pos == L'(') {
@@ -167,12 +165,9 @@ static int parse_util_locate_cmdsub(const wchar_t *in, const wchar_t **begin, co
                     }
                 }
             }
-            is_token_begin = is_token_delimiter(pos[0], is_first, pos[1]);
         } else {
             escaped = false;
-            is_token_begin = false;
         }
-        is_first = false;
     }

     syntax_error |= (paran_count < 0);
diff --git a/src/tokenizer.cpp b/src/tokenizer.cpp
index 861c57f61..9d48cdb7e 100644
--- a/src/tokenizer.cpp
+++ b/src/tokenizer.cpp
@@ -154,7 +154,6 @@ tok_t tokenizer_t::read_string() {
     int slice_offset = 0;
     const wchar_t *const buff_start = this->token_cursor;
     bool is_first = true;
-    bool is_token_begin = true;

     auto process_opening_quote = [&](wchar_t quote) -> const wchar_t * {
         const wchar_t *end = quote_end(this->token_cursor, quote);
@@ -193,7 +192,7 @@ tok_t tokenizer_t::read_string() {
         // has been explicitly ignored (escaped).
         else if (c == L'\\') {
             mode |= tok_modes::char_escape;
-        } else if (c == L'#' && is_token_begin) {
+        } else if (is_comment(this->token_cursor, buff_start)) {
             this->token_cursor = comment_end(this->token_cursor) - 1;
         } else if (c == L'(') {
             paran_offsets.push_back(this->token_cursor - this->start);
@@ -281,7 +280,6 @@ tok_t tokenizer_t::read_string() {
         FLOGF(error, msg.c_str(), c, c, int(mode_begin), int(mode));
 #endif

-        is_token_begin = is_token_delimiter(this->token_cursor[0], is_first, this->token_cursor[1]);
         is_first = false;
         this->token_cursor++;
     }
@@ -682,8 +680,12 @@ maybe_t<tok_t> tokenizer_t::next() {
     return result;
 }

-bool is_token_delimiter(wchar_t c, bool is_first, maybe_t<wchar_t> next) {
-    return c == L'(' || !tok_is_string_character(c, is_first, next);
+bool is_comment(const wchar_t *pos, const wchar_t *tok_start) {
+    if (*pos != L'#') return false;
+    if (pos == tok_start) return true;
+    wchar_t prev = *(pos - 1);
+    bool prev_is_first = (pos - 1) == tok_start;
+    return prev == L'(' || prev == L')' || !tok_is_string_character(prev, prev_is_first, *pos);
 }

 wcstring tok_first(const wcstring &str) {
diff --git a/src/tokenizer.h b/src/tokenizer.h
index fccff61db..5322a2d32 100644
--- a/src/tokenizer.h
+++ b/src/tokenizer.h
@@ -133,8 +133,8 @@ class tokenizer_t : noncopyable_t {
     }
 };

-/// Tests if this character can delimit tokens.
-bool is_token_delimiter(wchar_t c, bool is_first, maybe_t<wchar_t> next);
+/// Tests if the given position starts a comment (possible inside a command substitution).
+bool is_comment(const wchar_t *pos, const wchar_t *tok_start);

 /// Returns only the first token from the specified string. This is a convenience function, used to
 /// retrieve the first token of a string. This can be useful for error messages, etc. On failure,

shayaviv replied:
I like your solution! My only reservation is that `)` is interpreted as a delimiter - as I said in my [previous comment](https://github.com/fish-shell/fish-shell/pull/8695#discussion_r801608359), I think it shouldn't.

Escapes are not handled in `tok_is_string_character()` either - for example, it wouldn't know if `c = ';'` is escaped - it assumes that it isn't. So I think it is reasonable to burden the caller with escaping. We can also document that, perhaps like this:
```c
/// \c pos must not be an escaped character.
```

krobelus replied:
> Escapes are not handled in `tok_is_string_character()` either

Yeah, but that doesn't contradict that function's name.
Meanwhile `is_comment()` is a straight up lie if quoted/escaped, and I couldn't find a better name.

So I'd just merge as-is.

krobelus commented on src/tokenizer.cpp:
```diff
@@ -679,6 +682,10 @@ maybe_t<tok_t> tokenizer_t::next() {
 }
 
+bool is_token_delimiter(wchar_t c, bool is_first, maybe_t<wchar_t> next) {
+    return c == L'(' || !tok_is_string_character(c, is_first, next);
```
+bool is_token_delimiter(wchar_t c, bool is_first, maybe_t<wchar_t> next) {
+    return c == L'(' || !tok_is_string_character(c, is_first, next);

We also want closing parens, here is a test case:

    echo ()#)

shayaviv replied:
I see two different ways to interpret the test case you provided, and both yield the same result:
* expand empty list into an `echo` command, appending `#)` to each arg → a newline is printed
  - This is how I intended my code change to interpret this.
* expand empty list into an `echo` command, followed by a comment → a newline is printed
  - I assume this is how you meant we should interpret this.

Let's look at a different example:
```fish
echo (echo a)#b
```
I assume `a#b` should be printed in this case, since that is the current behavior. It also makes sense to me, since this interpretation is consistent with `echo a#b`. Do you agree?

krobelus replied:
Sorry, you're right, # is not a comment after )
My bad -- using ()#) as example was foolish, since it is trivially empty because of cartesion product expansion
